src/crimson/os/seastore/extent_placement_manager.h

```diff
@@ -538,16 +583,145 @@ class ExtentPlacementManager {
+        return create_tuple_impl<T>(std::make_index_sequence<N>{});
+      }
+
+      template<typename Tuple, typename Func>
+      static auto iterate_tuple_with_result(Tuple&& tuple, Func&& func) {
+        using TupleType = std::remove_reference_t<Tuple>;
+        using FirstTupleType = std::add_lvalue_reference_t<
+          std::tuple_element_t<0, TupleType>>;
+        using FuncReturnType = std::invoke_result_t<
+          Func, FirstTupleType, std::size_t>;
+        constexpr std::size_t tuple_size = std::tuple_size_v<TupleType>;
+
+        return []<std::size_t... I>(Tuple&& tuple, Func&& func,
+                                    std::index_sequence<I...>) {
+          auto res = create_tuple<FuncReturnType, tuple_size>();
+          ([&func, &tuple, &res]() {
+            std::get<I>(res) = func(std::get<I>(std::forward<Tuple>(tuple)), I);
+          }(), ...);
+          return res;
+        }(std::forward<Tuple>(tuple), std::forward<Func>(func),
+          std::make_index_sequence<tuple_size>{});
+      }
+
+      template<typename Tuple, typename Func>
+      static void iterate_tuple_without_result(Tuple&& tuple, Func&& func) {
+        using TupleType = std::remove_reference_t<Tuple>;
+        constexpr std::size_t tuple_size = std::tuple_size_v<TupleType>;
+
+        return []<std::size_t... I>(Tuple&& tuple, Func&& func,
+                                    std::index_sequence<I...>) {
+          (func(std::get<I>(std::forward<Tuple>(tuple)), I), ...);
+        }(std::forward<Tuple>(tuple), std::forward<Func>(func),
+          std::make_index_sequence<tuple_size>{});
+      }
+    };
+
+    template<typename... T>
+    auto try_reserve_impl(T&&... t) {
+      auto handler_tuple = std::make_tuple(create_reserve_handler(t)...);
+      std::set<std::size_t> failed_index;
+      auto res = details::iterate_tuple_with_result(
+        handler_tuple,
+        [&failed_index](auto &h, std::size_t idx) {
+          auto success = h->try_reserve();
+          if (!success) {
+            failed_index.insert(idx);
+          }
+          return success;
+        }
+      );
+
+      if (!failed_index.empty()) {
+        details::iterate_tuple_without_result(
+          handler_tuple,
+          [&failed_index](auto &h, std::size_t idx) {
+            if (!failed_index.contains(idx)) {
+              h->abort_reserve();
+            }
+          }
+        );
+      }
+
+      return res;
+    }
+
+    struct reserve_io_result_t {
       bool reserve_inline_success = true;
-      bool reserve_ool_success = true;
+      bool reserve_main_ool_success = true;
+
+      reserve_io_result_t(std::tuple<bool, bool> t)
+        : reserve_inline_success(std::get<0>(t)),
+          reserve_main_ool_success(std::get<1>(t)) {}
 
       bool is_successful() const {
-        return reserve_inline_success && reserve_ool_success;
-      }
+        return reserve_inline_success &&
+          reserve_main_ool_success;
+      };
     };
```

Comment:
How about grouping and moving the reservation primitives out of the ExtentPlacementManager class, to make the EPM class look simpler?